File: cs/Pragmastat.Tests/ShiftOverflowTests.cs
using Pragmastat;

namespace Pragmastat.Tests;

// Regression: shift search bounds x[0]-y[n-1] and x[m-1]-y[0] can overflow to
// +-Infinity on extreme finite input, turning the midpoint into NaN and returning
// +-Infinity instead of the true finite shift.
public class ShiftOverflowTests
{
  [Fact]
  public void ShiftDoesNotOverflowOnExtremeFiniteInput()
  {
    const double max = double.MaxValue;
    // Bitwise: the claim is the exact finite shift, and the first case is a difference of
    // symmetric samples, where +0.0 and -0.0 are both reachable and == cannot tell them apart.
    BitwiseAssert.Equal(0.0, Toolkit.Shift(new[] { -max, max }, new[] { -max, max }, assumeSorted: true),
      "Shift([-max, max], [-max, max])");
    BitwiseAssert.Equal(max, Toolkit.Shift(new[] { 0.0, max }, new[] { -max, 0.0 }, assumeSorted: true),
      "Shift([0, max], [-max, 0])");
  }
}

File: rs/pragmastat/tests/shift_overflow_tests.rs
//! Regression: shift search bounds x[0]-y[n-1] and x[m-1]-y[0] can overflow to
//! -/+inf on extreme finite input, turning the midpoint into NaN and returning
//! +-inf instead of the true finite shift.
//!
//! Both expectations are exact, so both compare payloads. `assert_eq!` would not:
//! the symmetric case asserts a zero, and `-0.0 == 0.0` holds, so a shift that
//! came back with the wrong sign of zero would be reported as a pass by the very
//! test written to pin the value.

mod conformance;

use conformance::assert_bits_eq;
use pragmastat::estimators::raw;

#[test]
fn shift_symmetric_extremes() {
    let max = f64::MAX;
    let got = raw::shift(&[-max, max], &[-max, max], true).unwrap();
    assert_bits_eq("shift of symmetric extremes", got, 0.0);
}

#[test]
fn shift_one_sided_extremes() {
    let max = f64::MAX;
    let got = raw::shift(&[0.0, max], &[-max, 0.0], true).unwrap();
    assert_bits_eq("shift of one-sided extremes", got, max);
}
